src/processor/operator/aggregate/aggregate_hash_table.cpp

```diff
@@ -804,9 +792,35 @@ uint64_t PartitioningAggregateHashTable::append(const std::vector<ValueVector*>&
     return numFlatTuples;
 }
 
-void PartitioningAggregateHashTable::mergeAll() {
-    partitioningData->appendTuples(*factorizedTable,
-        tableSchema.getColOffset(tableSchema.getNumColumns() - 1));
+bool hasSpace(uint64_t startingNumTuples, uint64_t newNumTuples, uint64_t currentMax) {
+    return startingNumTuples + newNumTuples > currentMax ||
+           static_cast<double>(startingNumTuples) + newNumTuples >
+               static_cast<double>(currentMax) / DEFAULT_HT_LOAD_FACTOR;
+}
+
+void PartitioningAggregateHashTable::mergeIfFull(uint64_t tuplesToAdd, bool mergeAll) {
+    if (mergeAll || hasSpace(factorizedTable->getNumTuples(), tuplesToAdd, maxNumHashSlots)) {
```

Comment:
I would refactor this into a function and reuse it at L846 too.

Reply:
I've simplified it a little. I'm not sure if it helps to pass `mergeAll` into that function, but the rest I could simply a bit.